src/Bicep.LangServer/BicepCompilationManager.cs

```diff
@@ -17,72 +21,146 @@ namespace Bicep.LanguageServer
         }
 
-        public void CloseCompilation(DocumentUri uri)
+        public void UpsertCompilation(DocumentUri uri, int? version, string fileContents)
         {
-            // remove the active compilation
-            this.activeContexts.TryRemove(uri, out _);
+            var newSyntaxTree = this.provider.BuildSyntaxTree(uri, fileContents);
+            var firstChanges = workspace.UpsertSyntaxTrees(newSyntaxTree.AsEnumerable());
+            var secondChanges = UpdateCompilationInternal(uri, version);
+            
+            var addedTrees = firstChanges.added.Concat(secondChanges.added);
+            var removedTrees = firstChanges.removed.Concat(secondChanges.removed);
 
-            // clear diagnostics for the file
+            foreach (var (entrypointUri, context) in activeContexts)
+            {
+                if (removedTrees.Any(x => context.Compilation.SyntaxTreeGrouping.SyntaxTrees.Contains(x)))
+                {
+                    UpdateCompilationInternal(entrypointUri, null);
+                }
+            }
+        }
+
+        public void CloseCompilation(DocumentUri documentUri)
+        {
+            // close and clear diagnostics for the file
             // if upsert failed to create a compilation due to a fatal error, we still need to clean up the diagnostics
-            PublishDocumentDiagnostics(uri, 0, Enumerable.Empty<OmniSharp.Extensions.LanguageServer.Protocol.Models.Diagnostic>());
+            CloseCompilationInternal(documentUri, 0, Enumerable.Empty<OmniSharp.Extensions.LanguageServer.Protocol.Models.Diagnostic>());
+        }
+
+        public void HandleFileChanges(IEnumerable<FileEvent> fileEvents)
+        {
+            var removedTrees = new HashSet<SyntaxTree>();
+            foreach (var change in fileEvents.Where(x => x.Type == FileChangeType.Changed || x.Type == FileChangeType.Deleted))
+            {
+                if (activeContexts.ContainsKey(change.Uri))
+                {
+                    // We should expect an explicit request to update this file if it's open. Otherwise we may
+                    // overwrite the 'dirty' copy of the file with a clean one.
+                    continue;
+                }
+
+                // We treat both updates and deletes as 'removes' to force the new SyntaxTree to be reloaded from disk
+                if (workspace.TryGetSyntaxTree(change.Uri.GetFileSystemPath(), out var removedTree))
+                {
+                    removedTrees.Add(removedTree);
+                }
+                else if (change.Type == FileChangeType.Deleted)
+                {
+                    // If we don't know definitively that we're deleting a file, we have to assume it's a directory; the file system watcher does not give us any information to differentiate reliably.
+                    // We could possibly assume that if the path ends in '.bicep', we've got a file, but this would discount directories ending in '.bicep', however unlikely.
```

Comment:
Totally agree with that. People will do unusual things in their directory structures.